File: src/weaponsprocurement/gui/StockReviewExpansionState.java
package weaponsprocurement.gui;

import weaponsprocurement.core.StockCategory;
import weaponsprocurement.core.StockItemType;

import java.util.EnumMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

final class StockReviewExpansionState {
    private final Map<StockCategory, Boolean> expanded = new EnumMap<StockCategory, Boolean>(StockCategory.class);
    private final Map<StockItemType, Boolean> expandedItemTypes = new EnumMap<StockItemType, Boolean>(StockItemType.class);
    private final Map<StockItemType, Map<StockCategory, Boolean>> expandedByItemType =
            new EnumMap<StockItemType, Map<StockCategory, Boolean>>(StockItemType.class);
    private final Map<StockReviewTradeGroup, Boolean> expandedTradeGroups = new EnumMap<StockReviewTradeGroup, Boolean>(StockReviewTradeGroup.class);
    private final Set<String> expandedItems = new HashSet<String>();

    StockReviewExpansionState() {
        expanded.put(StockCategory.NO_STOCK, Boolean.FALSE);
        expanded.put(StockCategory.INSUFFICIENT, Boolean.FALSE);
        expanded.put(StockCategory.SUFFICIENT, Boolean.FALSE);
        expandedItemTypes.put(StockItemType.WEAPON, Boolean.TRUE);
        expandedItemTypes.put(StockItemType.WING, Boolean.TRUE);
        initializeItemCategoryExpansion();
        expandedTradeGroups.put(StockReviewTradeGroup.BUYING, Boolean.FALSE);
        expandedTradeGroups.put(StockReviewTradeGroup.SELLING, Boolean.FALSE);
    }

    StockReviewExpansionState(StockReviewExpansionState source) {
        expanded.putAll(source.expanded);
        expandedItemTypes.putAll(source.expandedItemTypes);
        copyItemCategoryExpansion(source.expandedByItemType);
        expandedTradeGroups.putAll(source.expandedTradeGroups);
        expandedItems.addAll(source.expandedItems);
    }

    boolean isExpanded(StockCategory category) {
        Boolean value = expanded.get(category);
        return value != null && value.booleanValue();
    }

    void toggle(StockCategory category) {
        expanded.put(category, Boolean.valueOf(!isExpanded(category)));
    }

    boolean isExpanded(StockItemType itemType, StockCategory category) {
        Map<StockCategory, Boolean> byCategory = expandedByItemType.get(itemType);
        Boolean value = byCategory == null ? null : byCategory.get(category);
        return value != null && value.booleanValue();
    }

    void toggle(StockItemType itemType, StockCategory category) {
        if (itemType == null) {
            toggle(category);
            return;
        }
        Map<StockCategory, Boolean> byCategory = expandedByItemType.get(itemType);
        if (byCategory == null) {
            byCategory = new EnumMap<StockCategory, Boolean>(StockCategory.class);
            expandedByItemType.put(itemType, byCategory);
        }
        byCategory.put(category, Boolean.valueOf(!isExpanded(itemType, category)));
    }

    boolean isExpanded(StockItemType itemType) {
        Boolean value = expandedItemTypes.get(itemType);
        return value != null && value.booleanValue();
    }

    void toggle(StockItemType itemType) {
        expandedItemTypes.put(itemType, Boolean.valueOf(!isExpanded(itemType)));
    }

    boolean isExpanded(StockReviewTradeGroup tradeGroup) {
        Boolean value = expandedTradeGroups.get(tradeGroup);
        return value != null && value.booleanValue();
    }

    void toggle(StockReviewTradeGroup tradeGroup) {
        expandedTradeGroups.put(tradeGroup, Boolean.valueOf(!isExpanded(tradeGroup)));
    }

    void setExpanded(StockReviewTradeGroup tradeGroup, boolean value) {
        if (tradeGroup != null) {
            expandedTradeGroups.put(tradeGroup, Boolean.valueOf(value));
        }
    }

    boolean isItemExpanded(String itemKey) {
        return expandedItems.contains(itemKey);
    }

    void toggleItem(String itemKey) {
        if (itemKey == null || itemKey.isEmpty()) {
            return;
        }
        if (expandedItems.contains(itemKey)) {
            expandedItems.remove(itemKey);
        } else {
            expandedItems.add(itemKey);
        }
    }

    private void initializeItemCategoryExpansion() {
        for (StockItemType itemType : StockItemType.values()) {
            Map<StockCategory, Boolean> byCategory = new EnumMap<StockCategory, Boolean>(StockCategory.class);
            for (StockCategory category : StockCategory.values()) {
                byCategory.put(category, Boolean.FALSE);
            }
            expandedByItemType.put(itemType, byCategory);
        }
    }

    private void copyItemCategoryExpansion(Map<StockItemType, Map<StockCategory, Boolean>> source) {
        if (source == null || source.isEmpty()) {
            initializeItemCategoryExpansion();
            return;
        }
        for (StockItemType itemType : StockItemType.values()) {
            Map<StockCategory, Boolean> sourceByCategory = source.get(itemType);
            Map<StockCategory, Boolean> byCategory = new EnumMap<StockCategory, Boolean>(StockCategory.class);
            for (StockCategory category : StockCategory.values()) {
                Boolean value = sourceByCategory == null ? null : sourceByCategory.get(category);
                byCategory.put(category, Boolean.valueOf(value != null && value.booleanValue()));
            }
            expandedByItemType.put(itemType, byCategory);
        }
    }
}

File: src/main/kotlin/weaponsprocurement/ui/stockreview/state/StockReviewExpansionState.kt
package weaponsprocurement.ui.stockreview.state

import weaponsprocurement.ui.stockreview.trade.StockReviewTradeGroup
import weaponsprocurement.stock.item.StockCategory
import weaponsprocurement.stock.item.StockItemType
import java.util.EnumMap
import java.util.HashSet

class StockReviewExpansionState {
    private val expanded: MutableMap<StockCategory, Boolean> = EnumMap(StockCategory::class.java)
    private val expandedItemTypes: MutableMap<StockItemType, Boolean> = EnumMap(StockItemType::class.java)
    private val expandedByItemType: MutableMap<StockItemType, MutableMap<StockCategory, Boolean>> =
        EnumMap(StockItemType::class.java)
    private val expandedTradeGroups: MutableMap<StockReviewTradeGroup, Boolean> = EnumMap(StockReviewTradeGroup::class.java)
    private val expandedItems: MutableSet<String> = HashSet()

    constructor() {
        expanded[StockCategory.NO_STOCK] = false
        expanded[StockCategory.INSUFFICIENT] = false
        expanded[StockCategory.SUFFICIENT] = false
        expandedItemTypes[StockItemType.WEAPON] = true
        expandedItemTypes[StockItemType.WING] = true
        initializeItemCategoryExpansion()
        expandedTradeGroups[StockReviewTradeGroup.BUYING] = false
        expandedTradeGroups[StockReviewTradeGroup.SELLING] = false
    }

    constructor(source: StockReviewExpansionState) {
        expanded.putAll(source.expanded)
        expandedItemTypes.putAll(source.expandedItemTypes)
        copyItemCategoryExpansion(source.expandedByItemType)
        expandedTradeGroups.putAll(source.expandedTradeGroups)
        expandedItems.addAll(source.expandedItems)
    }

    fun isExpanded(category: StockCategory?): Boolean = expanded[category] == true

    fun toggle(category: StockCategory?): Boolean {
        if (category == null) return false
        expanded[category] = !isExpanded(category)
        return true
    }

    fun isExpanded(itemType: StockItemType?, category: StockCategory?): Boolean {
        val byCategory = expandedByItemType[itemType]
        return byCategory?.get(category) == true
    }

    fun toggle(itemType: StockItemType?, category: StockCategory?): Boolean {
        if (itemType == null) {
            return toggle(category)
        }
        if (category == null) {
            return false
        }
        val byCategory = expandedByItemType.getOrPut(itemType) { EnumMap(StockCategory::class.java) }
        byCategory[category] = !isExpanded(itemType, category)
        return true
    }

    fun isExpanded(itemType: StockItemType?): Boolean = expandedItemTypes[itemType] == true

    fun toggle(itemType: StockItemType?): Boolean {
        if (itemType == null) return false
        expandedItemTypes[itemType] = !isExpanded(itemType)
        return true
    }

    fun isExpanded(tradeGroup: StockReviewTradeGroup?): Boolean = expandedTradeGroups[tradeGroup] == true

    fun toggle(tradeGroup: StockReviewTradeGroup?): Boolean {
        if (tradeGroup == null) return false
        expandedTradeGroups[tradeGroup] = !isExpanded(tradeGroup)
        return true
    }

    fun setExpanded(tradeGroup: StockReviewTradeGroup?, value: Boolean): Boolean {
        if (tradeGroup == null || isExpanded(tradeGroup) == value) return false
        expandedTradeGroups[tradeGroup] = value
        return true
    }

    fun isItemExpanded(itemKey: String?): Boolean = expandedItems.contains(itemKey)

    fun toggleItem(itemKey: String?): Boolean {
        if (itemKey.isNullOrEmpty()) {
            return false
        }
        if (expandedItems.contains(itemKey)) {
            expandedItems.remove(itemKey)
        } else {
            expandedItems.add(itemKey)
        }
        return true
    }

    private fun initializeItemCategoryExpansion() {
        for (itemType in StockItemType.values()) {
            val byCategory: MutableMap<StockCategory, Boolean> = EnumMap(StockCategory::class.java)
            for (category in StockCategory.values()) {
                byCategory[category] = false
            }
            expandedByItemType[itemType] = byCategory
        }
    }

    private fun copyItemCategoryExpansion(source: Map<StockItemType, Map<StockCategory, Boolean>>?) {
        if (source.isNullOrEmpty()) {
            initializeItemCategoryExpansion()
            return
        }
        for (itemType in StockItemType.values()) {
            val sourceByCategory = source[itemType]
            val byCategory: MutableMap<StockCategory, Boolean> = EnumMap(StockCategory::class.java)
            for (category in StockCategory.values()) {
                byCategory[category] = sourceByCategory?.get(category) == true
            }
            expandedByItemType[itemType] = byCategory
        }
    }
}
